### PR TITLE
Add workflow for SARS-CoV-2 variant analysis

### PR DESCRIPTION
**Pull request by pvanheus**
This is a workflow for Illumina SARS-CoV-2 variant analysis with variant calling and consensus construction using iVar and lineage/clade assignment using pangolin and nextclade.

### CODE REVIEW

**Comment by bgruening** [comment]
@pvanheus can you add some more metadata please. E.g. a changelog file and dockstore.yml?

image

**Comment by bwlang** [comment]
@pvanheus I'm excited about this one... let me know how I can help.

**Comment by pvanheus** [comment]
> @pvanheus I'm excited about this one... let me know how I can help.

Thanks for your interest! I'm also working on [this tutorial](https://github.com/galaxyproject/training-material/pull/2633).

As [you](https://github.com/andersen-lab/ivar/issues/90) and others have noted, iVar is not without its issues. Expanding the set of test data to illustrate where problems might be would be useful. There have been a variety of approaches to addressing these, e.g. [this workflow](https://github.com/galaxyproject/iwc/tree/main/workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-variant-calling) from @wm75 and [this from covid-19-signal](https://github.com/jaleezyy/covid-19-signal/blob/master/Snakefile#L630) that uses Freebayes in addition to iVar. I think its worth getting this workflow in its current form out there so long though since this approach is in wide use worldwide.

**Comment by pvanheus** [comment]
Workflow testing is currently failing because `nextflow` is not available on usegalaxy.org. Waiting on pull requests from @mvdbeek to be merged to resolve this.

File: workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga
Comment (mvdbeek) [inline workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:1035]: The latest version seems to be 0.14.4+galaxy0, any reason not to use that one ?
```suggestion
            "content_id": "toolshed.g2.bx.psu.edu/repos/iuc/nextclade/nextclade/0.14.4+galaxy0",
```

File: workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga
Comment (mvdbeek) [inline workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:1072]:
```suggestion
                "changeset_revision": "3b75d5285aff",
```

File: workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga
Comment (mvdbeek) [inline workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:1078]:
```suggestion
            "tool_version": "0.14.4+galaxy0",
```

File: workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga
Comment (mvdbeek) [inline workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:1070]:
```suggestion
            "tool_id": "toolshed.g2.bx.psu.edu/repos/iuc/nextclade/nextclade/0.14.4+galaxy0",
```

File: workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga
Comment (mvdbeek) [inline workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:821]:
```suggestion
            "content_id": "toolshed.g2.bx.psu.edu/repos/iuc/multiqc/multiqc/1.9+galaxy1",
```

File: workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga
Comment (mvdbeek) [inline workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:884]:
```suggestion
            "tool_id": "toolshed.g2.bx.psu.edu/repos/iuc/multiqc/multiqc/1.9+galaxy1",
```

File: workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga
Comment (mvdbeek) [inline workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:886]:
```suggestion
                "changeset_revision": "75c93c70d094",
```

File: workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga
Comment (mvdbeek) [inline workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:892]:
```suggestion
            "tool_version": "1.9+galaxy1",
```

File: workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga
Comment (mvdbeek) [inline workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:891]:
```suggestion
            "tool_state": "{\"comment\": \"\", \"export\": \"false\", \"flat\": \"false\", \"results\": [{\"__index__\": 0, \"software_cond\": {\"software\": \"fastp\", \"__current_case__\": 7, \"input\": {\"__class__\": \"RuntimeValue\"}}}, {\"__index__\": 1, \"software_cond\": {\"software\": \"samtools\", \"__current_case__\": 23, \"output\": [{\"__index__\": 0, \"type\": {\"type\": \"stats\", \"__current_case__\": 0, \"input\": {\"__class__\": \"RuntimeValue\"}}}]}}, {\"__index__\": 2, \"software_cond\": {\"software\": \"qualimap\", \"__current_case__\": 19, \"input\": {\"__class__\": \"RuntimeValue\"}}}], \"saveLog\": \"false\", \"title\": \"\", \"__page__\": null, \"__rerun_remap_job_id__\": null}",
```

**Comment by mvdbeek** [comment]
@pvanheus it's green now, but I've suggested updates of nextclade and multiqc. Let me know if you want to merge as is and then update, or update right away

File: workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga
Comment (drosofff) [inline workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:459]: Hi @pvanheus there is also a newer version for this tool: qualimap_bamqc/2.2.2d+galaxy3

File: workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga
Comment (wm75) [inline workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:300]: Can you use samtools_view - toolshed.g2.bx.psu.edu/repos/iuc/samtools_view/samtools_view/1.9+galaxy3 - instead of this outdated tool please?

File: workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga
Comment (wm75) [inline workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:451]: You may want to consider "filter_by": "yes" here to make use of ivar trim's new `-f` option, which makes use of knowledge about expected amplicons to discard reads that do not fall within amplicon boundaries.

The downside is that you'll require one more input dataset, but it seems to be worth it -> https://github.com/andersen-lab/ivar/issues/75

Comment (wm75) [reply workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:451]: This is one of the improvements in v0.3 of (https://github.com/galaxyproject/iwc/tree/main/workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-variant-calling)

File: workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga
Comment (wm75) [inline workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:708]: This step seems *very* questionable!

Just renaming the reference to have it recognized by SnpEff can mask input data issues. What if the reference used by the user isn't identical to NC_045512.2?

Comment (wm75) [reply workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:708]: My assumption is you do this to keep both ivar trim and SnpEff happy?
If that's the case you can just omit this step and directly use NC_045512.2 as input. ivar trim will change the primer bed file ref name on the fly for you.

File: workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga
Comment (wm75) [inline workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:985]: This one absolutely needs refactoring.

Running Pangolin on a collection of Fastas means independent jobs will all download the latest database from the web!
Please concatenate the Fastas, not the pangolin output.

Probably the same works for nextclade? Which would even save a step in the WF then cause you'd have to concatenate only once upstream?

File: workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga
Comment (wm75) [inline workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:794]: Can you find a better name here?

What this does is it will create datasets named "ivar consensus on data xxx Consensus" just like the previous step (i.e. the name will be misleading both in terms of the tool and the input dataset #). I know you're hiding the previous step's output collection, but if you ever need to debug one of the resulting histories, you'll regret this choice.

**Comment by wm75** [comment]
As a general comment, you may want to work a bit more on the dataset/collection names of the workflow outputs.

Screenshot from 2021-06-21 18-12-55

You chose to hide intermediate collections, which is arguably good from an end-user's perspective, but then you need to take care that the names of the remaining datasets confer usable information. In the screenshot above, for example, it would be extremely tedious to track the provenance of that collection 331. It may be better to just call it "Annotated variants" then.

Personally I'd remove some of your WF step labels (not step output labels) and use them as the output names. WF step labels make it harder, imo, to see what a workflow does cause you have to click all the boxes in the editor to know which tool is getting used there (just my own personal opinion though).

**Comment by wm75** [comment]
and finally a question: why are you using an ivar variants AF threshold of 0.7 for calling variants, but a consensus threshold of 0.5. Wouldn't that occassionally incorporate a variant base into your consensus that isn't even reported in the annotated vcf?

I would have expected equal values here or, if anything, a *higher* consensus threshold, i.e., call and annotate relatively aggressively, but only build the consensus based on what looks credible?

**Comment by mvdbeek** [comment]
> Personally I'd remove some of your WF step labels (not step output labels) and use them as the output names. WF step labels make it harder, imo, to see what a workflow does cause you have to click all the boxes in the editor to know which tool is getting used there (just my own personal opinion though).

That's a shortcoming of the editor that is relatively easy to fix. If the labels communicate the intent of the step we should keep it, but that is a very good point, we need to fix that in the editor. It's particularly painful for labelled subworkflows.

File: workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga
Comment (pvanheus) [reply workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:708]: The renaming is because MN908947.3 is often used as input and yes it does have a risk like you outlined.

File: workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga
Comment (pvanheus) [reply workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:794]: In my experience, this renames the outputs to have the name of the collection element - i.e. if the input is called `sample1` the output is `sample1` in the collection. The naming of the collection as a whole remains `"iv.ar consensus on data xxx Consensus`.

Comment (wm75) [reply workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:794]: The renaming applies to the collection and its datasets. It's the collection elements' element identifers that are passed through like you say. In other words when you're viewing the datasets in the context of the collection you will see the names you're expecting, but when you view them as individual datasets (e.g. when you unhide hidden elements) or try to edit their names, you will see the "newname" set here.
Anyway, that newname is misleading for the collection and its elements.

File: workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga
Comment (wm75) [reply workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:708]: Ok, then what about this compromise: You replace specifically only MN908947.3 with NC_045512.2, but do this up front on the user-provided fasta file?

Those two are base identical, correct?

Comment (bgruening) [reply workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:708]: Should we not explicitly state which input we assume, which reference genome, and not hack around potential user assumptions. This is a general question for all workflows under IWC.

Comment (wm75) [reply workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:708]: In general, I agree (need to check our existing WFs 😊 ). In the specific case though MN908947.3 is the ID of the record the refseq NC_045512.2 was generated from and the two are identical and used kind of interchangeably (the ARTIC networks primer scheme files, e.g., use MN908947.3 instead of the refseq ID). They are identical (I just reconfirmed) so there's a point to making the WF convenient to use.
I don't think I will add such an extra step to the covid19.galaxyproject.org WFs, but this one here really emphasizes end-user friendliness so maybe it's an appropriate thing to do here.
As I said, I would like to see the replacement done as the very first thing in the WF though. No changing of IDs half-way through the history.

**Comment by bwlang** [comment]
I've done some reworking of this so it can be used as a subworkflow by a top-level ONT and top-level Illumina workflow.
@pvanheus : does that appeal to you? should I PR that into your fork?

**Comment by pvanheus** [comment]
Thanks @mvdbeek @wm75 @bgruening @drosofff I have updated the workflow and incorporated your suggestions. So:
1. The latest versions of tools are used
2. The reference consensus is renamed at the start of the workflow (and the rename is made more strict to only apply to this one case)
3. The pangolin and nextclade are run on concatenated FASTA
4. Each collection tool output does not use `on #{input}` in its labeling.

**Comment by pvanheus** [comment]
btw this workflow works when tested against https://galaxy.sanbi.ac.za

File: workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis-test.yml
Comment (wm75) [inline workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis-test.yml:13]: This file is needed for the -f option of ivar trim. You should probably make use of this option also in the training material, add the file to zenodo record 4944064, and get all test data from the same source then.

**Comment by wm75** [comment]
cosmetic change: The WF annotation is
`"annotation": "Find &amp; annotate variants in SARS-CoV-2 samples sequenced with Illumina ARTICv3",`
currently. Don't know how the `&amp;` slipped in there. Maybe the WF editor doesn't like `&`?

File: workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga
Comment (wm75) [inline workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:988]: Just out of interest: @mvdbeek is this tool or "Collapse collection" better suited for concatenating the elements of a collection?

Comment (mvdbeek) [reply workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:988]: tp_cat is simpler and has fewer options. It's a thin wrapper around cat, while `Collapse collection` has options to add the element_identifier and to keep one header line. Collapse collection` would probably reach a limit in the length of the generated command line for very big collections (https://github.com/bgruening/galaxytools/pull/1135) ... which could be fixed of course.

So for this use tp_cat is probably better.

**Comment by pvanheus** [comment]
> I've done some reworking of this so it can be used as a subworkflow by a top-level ONT and top-level Illumina workflow.
> @pvanheus : does that appeal to you? should I PR that into your fork?
> Also addressed Wolfgang's pangolin concern.

@bwlang said on chat that he will do his PR as an update after this one is merged.

**Comment by bwlang** [comment]
Any reason not to merge now?

File: workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis-test.yml
Comment (mvdbeek) [reply workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis-test.yml:13]: @pvanheus I think that comment still applies ?

File: workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga
Comment (mvdbeek) [inline workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:11]:
```suggestion
    "release": "0.1",
    "format-version": "0.1",
```

File: workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis-test.yml
Comment (pvanheus) [reply workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis-test.yml]: The Amplicon Primer Info file is now computed (by the `ivar trim` tool) so doesn't need to be provided by the user.

**Review by wm75 — APPROVED** [review]
Looks like a really mature first release version to me!
Congrats and thanks @pvanheus!

**Comment by drosofff** [comment]
Thank you @pvanheus ! If I did not misunderstood, I really really like the no-need-for-primer-bed feature !

**Pull request merged**
